### PR TITLE
[JAX] Regression tests for custom ops with jax.experimental.custom_partitioning

### PR DESCRIPTION
**Pull request by denera**
Included ops:
- `layernorm()` inherited from [Donglin Yang's fork](https://github.com/Young768/TransformerEngine/blob/main/tests/jax/test_ln_cp.py) (cc: @Young768) + `rmsnorm` type
- `softmax()` SCALED, SCALED_MASKED, SCALED_UPPER_TRIANG_MASKED
- `self_fused_attn()`, NO_BIAS, PRE_SCALE_BIAS, POST_SCALE_BIAS, NO_MASK, CAUSAL_MASK, PADDING_MASK
- `cross_fused_attn()`, NO_BIAS, NO_MASK, PADDING_MASK

Tested shardings:
- DP
- TP_COL
- TP_ROW (softmax only)
- DP_TP_COL
- DP_TP_ROW (softmax only)

### CODE REVIEW

File: tests/jax/test_custom_partitioning.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+from transformer_engine.jax.flax import extend_logical_axis_rules
+from transformer_engine.jax.sharding import get_dot_sharding_meta
+from transformer_engine.jax.sharding import get_elementwise_sharding_meta
+from transformer_engine.jax.sharding import get_fp8_meta_sharding_meta
+from transformer_engine.jax.sharding import global_shard_guard
+from transformer_engine.jax.sharding import infer_major_sharding_type
+from transformer_engine.jax.sharding import is_dp_enabled, is_tp_enabled
+from transformer_engine.jax.sharding import ShardingMeta, ShardingResource, ShardingType
+from transformer_engine.jax.sharding import get_elementwise_sharding_meta, extend_fsdp_sharding_meta
+from transformer_engine.jax.sharding_new import MeshResource
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:26]: We could removed `_new` here, since #472 is merged `*_new.py` to `*.py`.

File: tests/jax/test_custom_partitioning.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+from flax.linen import make_causal_mask
+from flax.linen import combine_masks
+from transformer_engine.jax.flax import extend_logical_axis_rules
+from transformer_engine.jax.sharding import get_dot_sharding_meta
+from transformer_engine.jax.sharding import get_elementwise_sharding_meta
+from transformer_engine.jax.sharding import get_fp8_meta_sharding_meta
+from transformer_engine.jax.sharding import global_shard_guard
+from transformer_engine.jax.sharding import infer_major_sharding_type
+from transformer_engine.jax.sharding import is_dp_enabled, is_tp_enabled
+from transformer_engine.jax.sharding import ShardingMeta, ShardingResource, ShardingType
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:24]: `ShardingMeta`, `ShardingResource` and `ShardingType` are deprecating. Will fully removed after the migration to CP is done.

File: tests/jax/test_custom_partitioning.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+def _get_sharding_spec(mesh_names, sharding_type):
+
+    if sharding_type is ShardingType.DP:
+        return P(mesh_names[0], None), P(None), P(None)
+    elif sharding_type is ShardingType.DP_TP_COL:
+        return P(mesh_names[0], mesh_names[1]), P(None), P(None)
+    else:
+        raise NotImplementedError
+
+def _get_sharding_resource(mesh_names, sharding_type):
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:43]: As above description, we are deprecating `ShardingType`, then this might break in the nearly feature.

File: tests/jax/test_custom_partitioning.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+                mask_ = make_decoder_mask(tokens)
+                graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+                pjitter = pjit(graded_f)
+                out = {}
+                hlo = pjitter.lower(x_, bias_, mask_).compile().as_text()
+                dic = count_collective(hlo)
+                assert dic==collective_ref, f"Expected number of collective is: {dic==collective_ref=}, but got {dic=}."
+
+class TestCPGenerator:
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:185]: Regarding softmax, it is expected 0 collective ops for all sharding. Since no weights need to reduce.

File: tests/jax/test_custom_partitioning.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+            pjitter = pjit(graded_f,
+                         in_shardings=[x_spec, gamma_spec, beta_spec],
+                        out_shardings=(None, (x_spec, gamma_spec, beta_spec),))
+                
+            hlo = pjitter.lower(x_, gamma, beta).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, gamma, beta)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:217]: Could we also add correctness checking of results?

File: tests/jax/test_custom_partitioning.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+                           in_shardings=[x_spec, mask_spec],
+                           out_shardings=(None, (x_spec),))
+                
+            hlo = pjitter.lower(x_, mask_).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, mask_)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
+
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:256]: Same as softmax.

File: tests/jax/test_custom_partitioning.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            graded_f = jax.value_and_grad(func, argnums=(0))
+            pjitter = pjit(graded_f,
+                           in_shardings=[x_spec, mask_spec],
+                           out_shardings=(None, (x_spec),))
+                
+            hlo = pjitter.lower(x_, mask_).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, mask_)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:254]: Same as `layernorm` to test correctness.

File: tests/jax/test_custom_partitioning.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            graded_f = jax.value_and_grad(func, argnums=(0, 1, 2))
+            pjitter = pjit(graded_f,
+                           in_shardings=[x_spec, bias_spec, mask_spec],
+                           out_shardings=(None, (x_spec, bias_spec, mask_spec),))
+                
+            hlo = pjitter.lower(x_, bias_, mask_).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, bias_, mask_)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
```
Comment (mingxu1067) [inline tests/jax/test_custom_partitioning.py:301]: Same as `layernorm` to test correctness.

**Comment by mingxu1067** [comment]
In summary, could we have following changes?
- To remove `Xmap` testing, since this is deprecating and would be removed in the nearly feature.
- To support different num of collective ops of each custom calls with different sharding patterns.
- To add correctness checking to guarantee the output values are expected.

**Comment by denera** [comment]
I'll expand the tests to check correctness against JAX-native `flax.linen` counterparts, but I'm not clear on the `Xmap` deprecation. 

Unless it's getting deprecated in JAX itself, it may be useful to keep it around at least internally so we can use it as a reference to check custom partitioning communication patterns. Otherwise I don't know what to use as reference in its place.

**Comment by nouiz** [comment]
> I'll expand the tests to check correctness against JAX-native `flax.linen` counterparts, but I'm not clear on the `Xmap` deprecation.
> 
> Unless it's getting deprecated in JAX itself, it may be useful to keep it around at least internally so we can use it as a reference to check custom partitioning communication patterns. Otherwise I don't know what to use as reference in its place.

I think we need to hardcode the expected collective we wants.
We won't keep xmap around, too much code and would be a big maintenance burden. If we hardcode the expectation, we won't need it.

File: tests/jax/test_custom_partitioning.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                         in_shardings=[x_spec, gamma_spec, beta_spec],
+                        out_shardings=(None, (x_spec, gamma_spec, beta_spec),))
+                
+            hlo = pjitter.lower(x_, gamma, beta).compile().as_text()
+            dic = count_collective(hlo)
+            test_l, test_grads = pjitter(x_, gamma, beta)
+            dic = count_collective(hlo)
+            assert dic==collective_ref, f"Expected number of collective is: {collective_ref=}, but got {dic=}."
+
+    @pytest.mark.parametrize('mesh_shape,mesh_names,sharding_type, collective_ref', MESH_CONFIG)
```
[the remark this answers is not in the document]
Comment (denera) [reply tests/jax/test_custom_partitioning.py:219]: This test for softmax with xla-based sharding is showing 1 all-reduce for me.

I would expect softmax to not have any collective ops only if the tensor is sharded so that each device has the entire tensor in the dimension aligned with the exponential sum. That doesn't seem to be happening here. If we want the softmax test to check for 0 collectives, we'll need to shard the tensor differently here, right?

[the remark this answers is not in the document]
Comment (nouiz) [reply tests/jax/test_custom_partitioning.py:219]: We need to test many different input sharding and for each, we need an hardcoded list of expected collective.

File: tests/jax/test_custom_partitioning.py (hunk 2 above)
Comment (denera) [reply tests/jax/test_custom_partitioning.py]: I think we should keep the old map-based sharding around for a while, encapsulated into its own submodule/file and selectively loaded instead of the new CP-based sharding if we detect an older Jax version that don't support the new custom ops. We can deprecate and remove it permanently down the line, assuming Jax doesn't change its mind again on sharding.

File: tests/jax/test_custom_ops_cpar.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
```
Comment (nouiz) [inline tests/jax/test_custom_ops_cpar.py:3]: Add a comment to tell what cpar mean?

File: tests/jax/test_custom_ops_xmap.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (nouiz) [inline tests/jax/test_custom_ops_xmap.py:1]: why different file for custom_p and xmap?

**Comment by nouiz** [comment]
@mingxu1067 can you also review?

**Comment by denera** [comment]
@mingxu1067 Quick note on xmap tests: I needed to run these ops and mesh configs with xmap shardings just to get the collective counts that are now hardcoded for the custom_partioning tests. I kept the code for that in the PR for now as regression tests for xmap in case there's a brief phase where we keep xmap alive with a deprecation warning instead of removing it immediately.

File: tests/jax/custom_ops_helper.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+
+import flax
+
+from transformer_engine.jax.sharding import ShardingType, ShardingResource
+try:
+    # temporary workaround to be removed after jax.experimental.custom_partitioning migration
+    from transformer_engine.jax.sharding import MeshResource
+except ImportError:
+    pytest.skip("Need working MeshResource implementation to test " +
+                "jax.experimental.custom_partitioning sharding.")
```
[the remark this answers is not in the document]
Comment (denera) [reply tests/jax/custom_ops_helper.py:22]: It's just a temporary workaround to make sure the custom partitioning tests don't run until we merge new migrated the custom ops.

Once we merge this PR, we can rebase the custom ops PR onto the new main and remove the try/catch statement before merging it.

File: tests/jax/sharding_configs.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class ShardingConfigs(object):
+    
+    def __init__(self, num_gpus=jax.device_count('gpu')):
+        super().__init__()
+        if num_gpus < 2:
+            raise ValueError(f"ShardingConfig: Need at least 2 GPUs, but got {num_gpus}.")
+        
+        self.device_count = min(num_gpus, 8)
```
[the remark this answers is not in the document]
Comment (denera) [reply tests/jax/sharding_configs.py:16]: I just saw that other Jax tests use len(jax.devices()) to check for sufficient GPUs. If that's a reliable count, I can switch to it in mine too.

File: tests/jax/test_custom_ops_xmap.py (hunk 11 above)
Comment (denera) [reply tests/jax/test_custom_ops_xmap.py:1]: I generally try to avoid going over 300-400 lines of code per Python file. I find that it improves readability and maintainability.

In this particular case, there's the additional convenience of simply deleting the independent xmap test file when we're ready to completely remove xmap sharding from TE. Easier and safer than stripping out xmap tests from a combined file.

File: tests/jax/test_custom_ops_cpar.py (hunk 10 above)
Comment (denera) [reply tests/jax/test_custom_ops_cpar.py:3]: Will do. Related to this, I am not attached to any file, class, method, etc. names here. Please let me know if you have any better, more concise and/or descriptive alternatives, especially if there's naming conventions in TE-JAX that I've missed or forgotten about.

File: tests/jax/sharding_configs.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        mesh_configs = [
+            ((self.device_count, 1),    ("dp", None), ShardingType.DP),
+            ((self.device_count, 1),    ("tp", None), ShardingType.TP_COL),
+            ((self.device_count, 1),    ("tp", None), ShardingType.TP_ROW),
+        ]
+        if self.device_count > 2:
```
[the remark this answers is not in the document]
Comment (denera) [reply tests/jax/sharding_configs.py:22]: It doesn't assume power of 2 GPUs, but it will always use even numbers to test DP_TP configurations because there's no other way to construct a 2D device mesh. That will leave an unused GPU when device count is 5 or 7. Device count of 6 fits cleanly here with a 3x2 grid even though it's not a power of two.

That said, I see that 3 GPUs are unaccounted for right now. It will try (and fail?) to test DP_TP with a 1x1 grid. I will modify the check to be >3 instead of >2 to avoid that.

File: tests/jax/sharding_configs.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (nouiz) [reply tests/jax/sharding_configs.py:16]: You can rely on len(jax.devices())

**Review by nouiz — APPROVED** [review]
